### PR TITLE
[4.x] fix(core): stringify shouldn't throw when toString returns null/undefined

### LINKED ISSUE #14948: stringify fails if token.toString() returns undefined
**I'm submitting a ...**  (check one with "x")
```
[x ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
We had a working Angular2 application running on Angular 2.2.0. 
Since we needed some new features we tried to upgrade to a newer Angular, but then our tests started failing. Seems to fail from Angular 2.3.0 and forward. 

The tests are UI tests creating our components in the TestBed and using mocks of the objects provided to our component. 

It seems that if the mocks, e.g. an object provided to a component during construction, returns undefined when calling .toString() on it - Angular throws an exception and stops working. 

The issue is only seen in test, as our production dependencies return something on toString(). 

I have attached a sample application exposing the error.
[StringifyBug.zip](https://github.com/angular/angular/files/820818/StringifyBug.zip)

Run 'ng test' to see the two tests. One test fakes a toString() result, the other dont. 

It seems to boil down to https://github.com/angular/angular/blob/master/modules/%40angular/facade/src/lang.ts#L102 where the result from toString() is used directly in indexOf call. 

Current workaround is to fake toString() on all our mocks so the test can run cleanly.

**Expected behavior**
The stringify method should be more defensive, and not break down if token.toString() returns null or undefined. 

**Minimal reproduction of the problem with instructions**
See the sample provided. 

- A component that takes something as constructor injection
- A test class, that using TestBed injects a fake object that returns null/undefined when calling toString() on it.
- Boom, it breaks.  

**What is the motivation / use case for changing the behavior?**
Stable framework, where it is possible to inject fake classes instead of e.g. Router, ActivatedRoute etc in test, without having to figure out internals.

**Please tell us about your environment:**
Windows 10, 64bit, npm, and Visual Studio Code.
Sample should be able to run with just "npm install", "npm run test"

* **Angular version:** 2.4.9 (reproducible from 2.3.0)

* **Browser:** [all]
 
* **Language:** [all]

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes #14948

### CODE REVIEW

File: packages/core/test/util_spec.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+import {stringify} from '../src/util';
+
+export function main() {
+  describe('stringify', () => {
+    it('should return string undefined when toString returns undefined', () => {
+      class Foo {
+        toString(): string { return undefined; }
+      }
+      expect(stringify(new Foo())).toBe('undefined');
```
Comment (vicb) [inline packages/core/test/util_spec.ts]: `expect(stringify({toString => undefined})).toBe('undefined');`

**Review by vicb — APPROVED** [review]
nit: could simplify the test

**Comment by DzmitryShylovich** [comment]
@vicb done. simplified the test + rebase

File: packages/core/test/util_spec.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {stringify} from '../src/util';
+
+export function main() {
+  describe('stringify', () => {
+    it('should return string undefined when toString returns undefined',
+       () => expect(stringify({toString: (): string => undefined})).toBe('undefined'));
```
Comment (vicb) [inline packages/core/test/util_spec.ts]: string -> any

**Comment by vicb** [comment]
@DzmitryShylovich thanks ! Could you please update the return type

**Comment by DzmitryShylovich** [comment]
@vicb done

**Pull request closed without merging**